homeassistant/components/icloud/__init__.py

```diff
@@ -1 +1,556 @@
+        username,
+        password,
+        account_name,
+        max_interval,
+        gps_accuracy_threshold
+    )
+    account.reset_account()
+
+    if account.api is None:
+        _LOGGER.error("No iCloud data added for account=%s", account_name)
+        return False
+
+    hass.data[DATA_ICLOUD][account.name] = account
+
+    for component in ICLOUD_COMPONENTS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    def play_sound(service):
+        """Play sound on the device."""
+        accountname = service.data.get(CONF_ACCOUNTNAME)
+        accountname = slugify(accountname.partition('@')[0])
+        devicename = service.data.get(ATTR_DEVICENAME)
+        devicename = slugify(devicename.replace(' ', '', 99))
+
+        hass.data[DATA_ICLOUD][accountname].devices[devicename].play_sound()
+
+    def display_message(service):
+        """Display a message on the device."""
+        accountname = service.data.get(CONF_ACCOUNTNAME)
+        accountname = slugify(accountname.partition('@')[0])
+        devicename = service.data.get(ATTR_DEVICENAME)
+        devicename = slugify(devicename.replace(' ', '', 99))
+        message = service.data.get(SERVICE_ATTR_LOST_DEVICE_MESSAGE)
+        sound = service.data.get(SERVICE_ATTR_LOST_DEVICE_SOUND, False)
+
+        hass.data[DATA_ICLOUD][accountname].devices[
+            devicename].display_message(
+                message,
+                sound)
+
+    def lost_device(service):
+        """Make the device in lost state."""
+        accountname = service.data.get(CONF_ACCOUNTNAME)
+        accountname = slugify(accountname.partition('@')[0])
+        devicename = service.data.get(ATTR_DEVICENAME)
+        devicename = slugify(devicename.replace(' ', '', 99))
+        number = service.data.get(SERVICE_ATTR_LOST_DEVICE_NUMBER)
+        message = service.data.get(SERVICE_ATTR_LOST_DEVICE_MESSAGE)
+
+        hass.data[DATA_ICLOUD][accountname].devices[devicename].lost_device(
+            number,
+            message)
+
+    def update(service):
+        """Call the update function of an iCloud account."""
+        accountname = service.data.get(CONF_ACCOUNTNAME)
+
+        if accountname is None:
+            for accountname, account in hass.data[DATA_ICLOUD].items():
+                account.keep_alive(utcnow())
+        else:
+            accountname = slugify(accountname.partition('@')[0])
+            hass.data[DATA_ICLOUD][accountname].keep_alive(utcnow())
+    
+    def reset_account(service):
+        """Reset an iCloud account."""
+        accountname = service.data.get(CONF_ACCOUNTNAME)
+
+        if accountname is None:
+            for accountname, account in hass.data[DATA_ICLOUD].items():
+                account.reset_account()
+        else:
+            accountname = slugify(accountname.partition('@')[0])
+            hass.data[DATA_ICLOUD][accountname].reset_account()
+
+    hass.services.register(
```

Comment:
Use the home assistant async api in async context. `hass.services.async_register`